server/src/main/java/org/opensearch/index/mapper/BooleanFieldMapper.java

```diff
@@ -257,9 +269,75 @@ public DocValueFormat docValueFormat(@Nullable String format, ZoneId timeZone) {
+            failIfNotIndexedAndNoDocValues();
+            if (!isSearchable()) {
+                return SortedNumericDocValuesField.newSlowExactQuery(name(), Values.TRUE.bytesEquals(indexedValueForSearch(value)) ? 1 : 0);
+            }
+            Query query = new TermQuery(new Term(name(), indexedValueForSearch(value)));
+            if (boost() != 1f) {
+                query = new BoostQuery(query, boost());
+            }
+            return query;
+        }
+
+        @Override
+        public Query termsQuery(List<?> values, QueryShardContext context) {
+            failIfNotIndexedAndNoDocValues();
+            int distinct = 0;
+            Set<?> distinctValues = new HashSet<>(values);
+            for (Object value : distinctValues) {
+                if (Values.TRUE.equals(indexedValueForSearch(value))) {
+                    distinct |= 2;
+                } else if (Values.FALSE.equals(indexedValueForSearch(value))) {
+                    distinct |= 1;
+                }
+                if (distinct == 3) {
+                    return this.existsQuery(context);
+                }
+            }
+            switch (distinct) {
+                case 1:
+                    return termQuery("false", context);
+                case 2:
+                    return termQuery("true", context);
+            }
+
+            return new MatchNoDocsQuery("Values did not contain True or False");
+        }
+
         @Override
         public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
-            failIfNotIndexed();
+            failIfNotIndexedAndNoDocValues();
+            if (isSearchable() && hasDocValues()) {
+                Query query = new TermRangeQuery(
+                    name(),
+                    lowerTerm == null ? null : indexedValueForSearch(lowerTerm),
+                    upperTerm == null ? null : indexedValueForSearch(upperTerm),
+                    includeLower,
+                    includeUpper
+                );
+                Query dvQuery = new TermRangeQuery(
+                    name(),
+                    lowerTerm == null ? null : indexedValueForSearch(lowerTerm),
+                    upperTerm == null ? null : indexedValueForSearch(upperTerm),
+                    includeLower,
+                    includeUpper,
+                    MultiTermQuery.DOC_VALUES_REWRITE
+                );
+                return new IndexOrDocValuesQuery(query, dvQuery);
+            }
+            if (hasDocValues()) {
+                return new TermRangeQuery(
+                    name(),
+                    lowerTerm == null ? null : indexedValueForSearch(lowerTerm),
+                    upperTerm == null ? null : indexedValueForSearch(upperTerm),
+                    includeLower,
+                    includeUpper,
+                    MultiTermQuery.DOC_VALUES_REWRITE
+                );
```

Comment:
I don't think these will work for doc values. The doc values are numeric, so a term-based doc value rewrite shouldn't work. (We don't have a yaml REST test for boolean range queries with doc values only, but if we did, I think they would fail with the wrong DV type.)

Based on the possible return values from `indexedValueForSearch`, there are only a few possible cases:

1. lowerTerm is `null`:
    1. Upper term is `null`. -> It's an `existsQuery`.
    2. Upper term is `Value.FALSE`. If `includeUpper`, then it's `termQuery("false", context);`, else it's `MatchNoDocs`.
    3. Upper term is `Value.TRUE`. If `includeUpper` then it's an `existsQuery`, else it's `termQuery("false", context);`
2. lowerTerm is `Value.FALSE`, and `includeLower`:
    1. If `upperTerm` is `null`, then `existsQuery`.
    2. If `upperTerm` is `Value.FALSE` and `includeUpper` then `termQuery("false")`
    3. If `upperTerm` is `Value.FALSE` and `includeUpper == false` then `MatchNoDocs`
    4. If `upperTerm` is `Value.TRUE` and `includeUpper` then `existQuery`.
    5. If `upperTerm` is `Value.TRUE` and `includeUpper==false` then `termQuery("false")`
3. lowerTerm is `Value.FALSE` and `includeLower == false`:
    1. If `upperTerm` is `null`, then `termQuery("true")`.
    2. If `upperTerm` is `Value.FALSE` then `MatchNoDocs` (regardless of `includeUpper`)
    3. If `upperTerm` is `Value.TRUE` and `includeUpper` then `existQuery`.
    4. If `upperTerm` is `Value.TRUE` and `includeUpper==false` then `termQuery("false")`
4. If lowerTerm is `Value.TRUE`:
    1. If `upperTerm` is `Value.TRUE` and `includeLower == includeUpper == true`, then it's `termQuery("true")`.
    2. If `upperTerm` is `null` and `includeLower == true`, then it's `termQuery("true")`.
    2. Else it's `MatchNoDocs`

I think that logic could be considerably simplified as:
1. If `lowerTerm == null`, then you rewrite it as `lowerTerm = Value.FALSE` and `includeLower = true`.
2. If `upperTerm == null`, then you rewrite it as `upperTerm = Value.TRUE` and `includeUpper = true`.
3. If `lowerTerm == upperTerm`, then it's `MatchNoDocs` if either `includeLower` or `includeUpper` is `false`. If they're both `true`, it will be a `termQuery`.
4. Now you've got the `lowerTerm != upperTerm` cases. 
    1. If `lowerTerm == VALUE.TRUE`, then it's `MatchNoDocs` (since the range is `TRUE,FALSE`, which is always empty, because it's backwards).
    2. So, you're left with the `FALSE,TRUE` cases and you need to look at the truth table for `includeLower` and `includeUpper`: `FF -> MatchNoDocs`, `TF -> termQuery("false")`, `FT -> termQuery("true")`, `TT -> existsQuery()`.
    
Easy! 😄

Reply:
You're right! not sure how I did not test the range query, but there are tests now. 

Essentially I think it's okay to offload this to the termQuery. I contemplated using the `SortedSetDocValues.newSlowRangeQuery` but I think termQuery is alright.